### DISCUSSION TITLE
bitShake Smartmeter Reader: Tasmota uses a "future" meter reading value on WebSend (extrapolated?)

### DISCUSSION THREAD

**Discussion by krza** [arendst/Tasmota#19684] [Support request] [answered]
Edit 23/10/10:
- Main root cause is the precision loss on converting from double to single.
- The workaround is to use `smls[-x]`. See my marked post below.
<hr>

Hi,

my energy meter sends new data all 2 seconds or something which is collected with a reader (see setup details at the bottom). Within the `>M` block of the script all values to be shown on the local web interface are defined. The `>S` block where the WebSend command is located refers to these lines through `sml[x]`. WebSend is performed all 15 seconds.

The local display of the values seems to be okay. They are updated with the mentioned 2 seconds rate (technically probably once a second). However, I hve an issue with the total meter reading in the WebSend term.

**Issue:**

The meter reading value `sml[1]` which is used for the WebSend command is not identical to the one which is correctly shown in the local display in the web browser:
- it is higher
- it is updated at a much lower rate

**Details:**

The reading is delivered with 4 decimal places, e.g. `51215.6196`. This is shown in the browser. But the WebSend command uses the value `51215.6211`, and this value does not change for 4 or 8 cycles, thus for exactly 1 or 2 minutes. This is the case at least if the power consumtopn is low and the value changes slowly. First I had the impression that the WebSend value is updated when the previously shown value was "overtaken" by the real value but this does not seem to be always true. In a specific case it took another full minute to get an update (which was higher than the actual value again).

The I added a higher load (about 1400 W) in order to increase the change rate, and this led to value updates within each 15 seconds cycle, all with that ominous offset though. After switching the load off te update rate resumed to once a minute or two.

It seems that the Tasmota script tries to extrapolate the value over time somehow and prevets "too" frequent updates at the same time. However, this does not seem to happen for the voltage and amps values. They are updated in each cycle if they change at all. This is another indicator that the extrapolation is only done for the specific `1-0:1.8.0` value because Tasmota "knows" its continuously and strongly increasing nature.

The problem is that this behavior makes the values unusable because they are firstly wrong (extrapolation) and secondly too slow resp. with a too low resolution (reduced cycle time).

Is there a simple fix for that? Maybe a configuration which switches that off?

Thanks a lot for any hints.

**Script:**
```
>D
ZSWH=0

>B
->sensor53 r

>S
ZSWH=sml[1]
if upsecs%15==0
then
=>WebSend [api.thingspeak.com] /update.josn?api_key=*&field1=%4sml[1]%&field2=%0sml[8]%&field3=%0sml[9]%&field4=%0sml[10]%&field5=%0sml[11]%&field6=%4ZSWH%&field7=%0sml[12]%
endif

>M 1
+1,5,o,0,9600,OBIS
1,1-0:1.8.0*255(@1,  Zählerstand, kWh, ZS,   4
1,=h <hr>
1,1-0:32.7.0*255(@1, Volt L1,     V,   V_L1, 1
1,1-0:52.7.0*255(@1, Volt L2,     V,   V_L2, 1
1,1-0:72.7.0*255(@1, Volt L3,     V,   V_L3, 1
1,=h <hr>
1,1-0:31.7.0*255(@1, Ampere L1,   A,   A_L1, 2
1,1-0:51.7.0*255(@1, Ampere L2,   A,   A_L2, 2
1,1-0:71.7.0*255(@1, Ampere L3,   A,   A_L3, 2
1,=h <hr>
1,1-0:21.7.0*255(@1, Watt L1,     W,   W_L1, 0
1,1-0:41.7.0*255(@1, Watt L2,     W,   W_L2, 0
1,1-0:61.7.0*255(@1, Watt L3,     W,   W_L3, 0
1,=h <hr>
1,=m 8+9+10      @1, Watt Gesamt, W,   real, 0
1,=h <hr>
1,=m 1*#10000    @1,     Wh Mul,  Wh,  WhM,  0 ; debug
1,1-0:1.8.0*255(@0.0001, Wh Div,  Wh,  WhD,  0 ; debug
#
```

**Setup:**
```
- Smart meter: Hager EHZ361W5 (OBIS)
- bitShake SmartMeterReader - WiFi Set
- ESP8266EX
- Tasmota 12.1.1, 2_7_4_9/2.2.2-dev(38a443e)
- Meter Values (d1):
  1-0:0.0.0     (meter s/n)
  1-0:1.8.0     (total meter reading)
  1-0:96.5.5    (82)
  0-0:96.1.255  (ID)
  1-0:32.7.0    (L1 V)
  1-0:52.7.0    (L2 V)
  1-0:72.7.0    (L3 V)
  1-0:31.7.0    (L1 A)
  1-0:51.7.0    (L2 A)
  1-0:71.7.0    (L3 A)
  1-0:21.7.0    (L1 W)
  1-0:41.7.0    (L2 W)
  1-0:61.7.0    (L3 W)
  1-0:96.50.0*x (ID)
```

**Comment by gemu2015** [comment]
sml driver works with double precision float values, while scripter works with single precision.
this results in imprecise sml[x] values when values with higher precision than 23 bit shall be displayed.

you can fetch the double precision value as a string with smls[-x]

```
>D
str=""
>S
str=smls[-1]
print %str%
```

will fetch a number string with full precision.

**Comment by krza** [comment]
You are right, at least the wrong number values are plausible then.

However the `-1` trick does not seem to work without a string conversion. Will have a look into it later the day, especially because it´s unclear why the script seems to use single and double precision depending on ... what? I thought the complete environmeht runs in single precision.

**Reply by gemu2015** [reply]
since SML has to handle very large values sometimes it generally uses double precision

in scripting double precision would be an overhead for most applications

i checked the fetch described above just now and it works like expected

it is a string not a number so you can't do any math with it but send it with websend

**Comment by krza** [comment]
Okay, thanks for the explanation. The sentence "all numbers are float" here in the scripting documentation in connection with a statement that "float" means "single" here had set me on the wrong track. I mean I did not x-check for a precision error because I thought that everything was single anyway. But again yes, if you have a 50k reading, the error is surprisingly high and destroys your digits behind the dot already.

Back to the point:

I still wonder why the following 2 lines in the scripting section work fine:
```
1,=m 1*#10000    @1,     Wh Mul,  Wh,  WhM,  0 ; debug
1,1-0:1.8.0*255(@0.0001, Wh Div,  Wh,  WhD,  0 ; debug
```
Both effectively do the same and multiply the value by 10.000 whereas the first line does a multiplication and the second one divides by 0.0001 instead. And the results are correct. Also the display of line 1 is correct:
```
1,1-0:1.8.0*255(@1,  Zählerstand, kWh, ZS,   4
```
So here no conversion to single precision seems to happen while it does in the `>S` section.

Coming back to your proposal. It does not work for me. I have tried several things. If I take your code as it is, the `str` variable is 0 (displayed as "0.00" in console resp. empty in WebSend string). If I change back to `sml[1]` or `sml[-1]` (thus without the "s"), it works again but in single precision and both with the same result.

For further testing I had added a few prints to the `>S` section with the following results:
```
Original value: 51220.8082   (displayed on the main menu page)

Console Out  Print Term    Required calculation to get the Out value
-----------  -----------   -----------------------------------------
51220.8086   %4sml[1]%     dbl2sng(original value)
512208064    %0sml[12]%    dbl2sng(original value * 10000) 
512208096    %0sml[13]%    dbl2sng(original value / dbl2sng(0.0001))
51220.8086   %4sml[-1]%    dbl2sng(original value)
51220.8086   %4sml[-12]%   dbl2sng(original value)
51220.8086   %4sml[-13]%   dbl2sng(original value)
```
So it is consistent wrt the double to single conversion between ... let´s say the `>M` section and the `>S` section. And in this specific case I would have thought that the minus sign in the `sml[-x]` term just refers to the original value before potential modifications within the lines in the `>M` section but does not seem to have an effect wrt the precision. See the values 12 and 13 for example.

**Reply by gemu2015** [reply]
all lines in the >M section are passed to the SML driver, so actually do not belong to scripter and are handled as double.
every other code in scripter defaults to single precision float.

smls[x] gets a string from SML driver.  positive indexes get a real string variable from SML (only 1 per meter allowed), while negative numbers get numeric values from SML numeric vars in double precision converted to a string.

as mentioned above i checked that this still works in the current release,  (what release are you using ?)

btw there is another possible solution by compiling your own version
by defining 
#define TS_FLOAT double 
in config override you may force scripter to use double floats and then your problem is also solved (this however uses more RAM and Flash)

**Reply by krza** [reply]
Okay, understood. Thanks a lot. In the regular documentation such details are often either kind of hidden or not available :)

So in principle it should work ... even without using a variable in between. But I had tested both. And again ... And now I get empty strings rather than 0.00 ... strange ... maybe I should restart the device? Such is said to be helpful for PCs at least ;)

My release was stated in my initial post (footer). But even if I would be able to compile somehting else I would not be able to flash it because I don´t have such equipment.

**Reply by krza** [reply]
No success. Restarting has not helped (as expected). So the `smls[x]` term does not work in my environment or there is some other blocking thing. Any further ideas?

**Reply by gemu2015** [reply]
your version is too old, you have to update to current version. this is easy via OTA.

**Reply by krza** [reply]
Okay, I see ... so 12.1.1 is too old and has to be upgraded to 13.1.0? I don´t know the overall architecture but have the "beware of the upgrade button" remarks in mind (in connection with the energy meter readers). Would that button in principle trigger an OTA upgrade, by the way? Could I also apply a _downloaded_ image or something to the ESP through OTA / Wifi? General question.

I have asked bitshake if an upgrade is feasible somehow.

**Reply by gemu2015** [reply]
i assume you are german ? then visit this page https://ottelo.jimdofree.com/stromzähler-auslesen-tasmota/

**Reply by krza** [reply]
How did you know ... ;)
> 1,1-0:1.8.0*255(@ 1,  **Zählerstand**, kWh, ZS,   4

Thanks for the link. Will have a look at it!

**Reply by sfromis** [reply]
You should not see suggestions to be careful to not just upgrade because a higher version numbers exist as a strong warning, merely to not see keeping current as something you should do for no other reason. When you run into difficulties, it is a normal thing that newer versions have improvements and fixes, but occasionally also breaking changes, bugs or just something which handles some situation differently. Still, when you're testing after the upgrade, just fine.

Using the upgrade feature in the web UI both allows you to do OTA from an external webserver, your own server, or through file upload.

**Reply by krza** [reply]
Hi sfromis,

in the specific case the upgrade would lead to a default load which does not contain the SML functionality. So doing that would destroy your function and even your script. So you should be prepared at least ;) And I would not go for an update if there is no specific reason. NTARS. But in this case there was at least the hope to change something wrt the string conversion.

And meanwhile I can at least confirm that the OTA upgrade to `13.0.0` via `13.1.0` has worked. A big thanks to the bitShake support. First step was to upgrade to a "minimal" variant of `13.1.0`, second step was to load a specific `13.0.0` image which contains the SML support again. So I am at least back to where I came from without any loss of functionality as it seems. Now I will test if the `smls[x]` stuff has changed ...

**Reply by sfromis** [reply]
You should obviously know which binary to upgrade to, what features it includes, and then you are not destroying anything.

**Reply by krza** [reply]
Yes. And this is what separates "simple users" from "power users" ;) In this case I was part of the first group because I did not even know Tasmota before. But now I am on the bridge between both groups if you like.

**Comment by krza** [comment]
Unfortunately it still does not work. Partly it´s even worse than before as it seems.

In principle `smls[x]` does something now but it´s still not the expected behavior. In addition to that the processing of the math lines seems to be slower than before. Without knowing the exact architecture it´s all stabbing around in the dark though ;)

Coming back to the "s": `smls[x]` returns an empty string, `smls[-x]` returns a value. However, latter is not the raw double value but a single value with 2 decimal places. Using `%4smls[-x]%` or `%1smls[-x]%` does not change anything wrt to the decimal places which confirms that it´s a string but no number. Using `sml[x]` or `sml[-x]` returns the same single value but as number.

Coming back to the following 2 lines again:
```
1,=m 1*#10000    @1,     Wh Mul,  Wh,  WhM,  0 ; debug 12
1,1-0:1.8.0*255(@0.0001, Wh Div,  Wh,  WhD,  0 ; debug 13
```
Both still show the correct (double) value in the Main Menu display as before. No change there. So I wonder where the conversion to single happens. Is the `>M` section running in double mode because it´s the SML part whereas the `>S` section is the scripting part which runs in single mode?

Second point: After the upgrade line 12 is "slow" now. Its update rate has decreased. Before both lines were always in SYNC, actually all 3 lines (1, 12, 13). But now only lines 1 and 13 are in sync but line 12 is late. And I don´t see a rule or something – sometimes it catches up after 1..2 cycles, sometimes it takes quite many cycles. All with a similar change rate of the original value.

If I change the math line as follows, by the way ...
```
1,=m 1-#51238    @1,     Wh Mul,  Wh,  WhM,  0 ; debug 12
```
thus if I replace the multiplication by 10000 by substracting the integer part of the value, I get a correct value in the WebSend term as well because the reduced value leads to a better precision. But firstly I would need to adapt this value on a regular basis in order to keep the resulting value small. And secondly there is still the other issue that the update rate of the WebSend term is much too slow for some reason. It runs all 15 seconds but repeats the same value for seveal times on a low change rate. No idea what the trigger is, so what the `>S` block is waiting for ;)

**Reply by gemu2015** [reply]
the default number of decimals is 2, you must expand this with the dp cmd before calling sml[-x] e.g. dp(8)
preceding the var with precision e.g. %8var% does not work for strings.

i don't know what you mean by slow down.  i can't see any slowing down im my setup

**Comment by krza** [comment, marked as answer]
Okay, I have a working solution now which I would like to post here to get issue closed and in case someone else can make use of the findings. Thanks again for all the help and patience ;)
```
>D
str=""                      ; string var for meter reading (actual)
str1=""                     ; string var for meter reading (-15 s)
str2=""                     ; string var for meter reading (-10 s)
str3=""                     ; string var for meter reading ( -5 s)
i=1                         ; loop counter
t:mytimer=0                 ; timer variable

>B
->sensor53 r

>S
dp(4)                       ; forces 4 decimal places on double to string conversion.
str=smls[-1]                ; converts raw meter reading (double) to string (e.g. "54321.1234").
if mytimer==0               ; check if timer has elapsed.
then
  mytimer=4.9               ; on 5 the timer is too slow but on 4.9 the rate is 5 seconds.
  switch i                  ; check loop counter.
  case 1
    str1="0."+sb(str -4 4)  ; preserves the meter reading 15 seconds before next WebSend.
  case 2
    str2="0."+sb(str -4 4)  ; preserves the meter reading 10 seconds before next WebSend.
  case 3
    str3="0."+sb(str -4 4)  ; preserves the meter reading  5 seconds before next WebSend.
  case 4
    =>WebSend [api.thingspeak.com] /update.josn?api_key=EJDANSSKFPCK4R77&field1=%str%&field2=%0sml[8]%&field3=%0sml[9]%&field4=%0sml[10]%&field5=%0sml[11]%&field6=%str1%&field7=%str2%&field8=%str3%
    i=0                     ; reset loop counter.
  ends
  i+=1                      ; increase loop counter.
endif
                            ; I had used "upsecs%15==0" for a 15 seconds loop before
                            ; (the free thingspeak broker accepts input with a rate
                            ; of at least 15 seconds only). However, it did not match
                            ; nicely, so I got 30 instead of 15 seconds gaps in the
                            ; online data quite often. So I changed it to 20 to be
                            ; on the safe side. But 20 was even worse than 15, so I
                            ; could not really see short load peaks/slopes. Usually
                            ; this is not much umportant but there are cases where
                            ; I would like to see the consumption changes a bit more
                            ; precisely. For that reason I collect "interim readings"
                            ; all 5 seconds, thus 3 of them between each WebSend.
                            ; This way I have the chance to increase the regular
                            ; 20 seconds rate to a 5 seconds rate for specific
                            ; analysis tasks. Of course the timing of the meter,
                            ; of the script and of WebSend are not very precise
                            ; and especially not aligned. But it´s good enough
                            ; and the "pseudo 5 seconds rate" was a nice gimmick
                            ; to play around with scripting and MQTT ;)
                            ;
                            ; The total consumption from line 11 below is not very
                            ; accurate, by the way. The calculation seems to be
                            ; delayed or something, at least it does not take the
                            ; values which are displayed in lines 8, 9 and 10 which
                            ; often leads to discrepancies if one of the Lx values
                            ; had just changed before (or afterwards?).

>M 1
+1,5,o,0,9600,OBIS
1,1-0:1.8.0*255(@1,  Zählerstand, kWh, ZS,   4 ;1
1,=h <hr>
1,1-0:32.7.0*255(@1, Volt L1,     V,   V_L1, 1 ;2
1,1-0:52.7.0*255(@1, Volt L2,     V,   V_L2, 1 ;3
1,1-0:72.7.0*255(@1, Volt L3,     V,   V_L3, 1 ;4
1,=h <hr>
1,1-0:31.7.0*255(@1, Ampere L1,   A,   A_L1, 2 ;5
1,1-0:51.7.0*255(@1, Ampere L2,   A,   A_L2, 2 ;6
1,1-0:71.7.0*255(@1, Ampere L3,   A,   A_L3, 2 ;7
1,=h <hr>
1,1-0:21.7.0*255(@1, Watt L1,     W,   W_L1, 0 ;8
1,1-0:41.7.0*255(@1, Watt L2,     W,   W_L2, 0 ;9
1,1-0:61.7.0*255(@1, Watt L3,     W,   W_L3, 0 ;10
1,=h <hr>
1,=m 8+9+10      @1, Watt Gesamt, W,   real, 0 ;11 (total consumption is not delivered by the meter)
1,=h <hr>
#
```
